hathor/nanocontracts/storage/contract_storage.py

```diff
@@ -0,0 +1,368 @@
+# Copyright 2023 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# XXX: avoid using `from __future__ import annotations` here because `make_dataclass_nc_type` doesn't support it
+
+import hashlib
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import TypeVar
+
+from hathor.conf.settings import HATHOR_TOKEN_UID
+from hathor.nanocontracts.nc_types import BytesNCType, NCType
+from hathor.nanocontracts.nc_types.dataclass_nc_type import make_dataclass_nc_type
+from hathor.nanocontracts.storage.maybedeleted_nc_type import MaybeDeletedNCType
+from hathor.nanocontracts.storage.patricia_trie import PatriciaTrie
+from hathor.nanocontracts.storage.token_proxy import TokenProxy
+from hathor.nanocontracts.storage.types import _NOT_PROVIDED, DeletedKey, DeletedKeyType
+from hathor.nanocontracts.types import BlueprintId, TokenUid, VertexId
+from hathor.serialization import Deserializer, Serializer
+
+T = TypeVar('T')
+D = TypeVar('D')
+
+_BYTES_NC_TYPE: NCType[bytes] = BytesNCType()
+
+
+class _Tag(Enum):
+    ATTR = b'\0'
+    BALANCE = b'\1'
+    METADATA = b'\2'
+
+
+class TrieKey(ABC):
+    @abstractmethod
+    def __bytes__(self) -> bytes:
+        raise NotImplementedError
+
+
+@dataclass(frozen=True, slots=True)
+class AttrKey(TrieKey):
+    nc_id: bytes
+    key: bytes
+
+    def __bytes__(self) -> bytes:
+        return _Tag.ATTR.value + hashlib.sha1(self.key).digest()
```

Comment:
Should we keep using sha1? In fact, do we need to hash at all? I guess so because it is what guarantee that the tree will be balanced.